text/0000-array-private-classes

```diff
@@ -0,0 +1,285 @@
+- Feature Name: array_private_classes
+- Start Date: 2022-01-08
+- RFC PR: (leave this empty)
+- Pony Issue: (leave this empty)
+
+# Summary
+
+Collection classes should not expose internal classes through iterators functions.
+
+Following information hiding design principle, the builtin classes of collection
+data structures must not be made visible through iterator functions like `ArrayKeys`,
+`ArrayValues` and `ArrayPairs`. These classes can be made private as they are
+only used as return types for `Array` functions `keys`, `values` and `pairs`.
+The return values for these functions are changed to the more general interface
+`Iterator`.
+
+A new interface `RewindableIterator` is defined to allow for rewindable iterators,
+like it is the case for `Array` `values`.
+
+This design principle is applied to the other collection classes that expose
+internals too like:
+
+* `List`
+* `Map`
+* persistent `Map`
+* `Vec`
+* persistent `Vec`
+* `Set`
+* `Itertools`
+
+This is a breaking change for collections' client code that use now internal
+classes but a search on Github repositories shows that the impact should be
+limited.
+
+# Motivation
+
+This change brings:
+
+- Applying the design principle of
+[hiding implementation details](https://en.wikipedia.org/wiki/Information_hiding)
+but offer a general and stable interface. Returning interfaces instead of concrete
+classes allows changing the implementation. Usually, one must return the most
+general type that fullfils the contract of the function (in the case of the
+functions discussed in this RFC, iteration).
+- Collections' functions `keys`, `values` and `pairs` definitions are made more
+general. Iterators implementation details are not public. Internal classes used
+by implementation like `*Keys`, `*Values` and `*Pairs` are now
+[opaque data types](https://en.wikipedia.org/wiki/Opaque_data_type). Generally,
+when using these collection classes, clients are not interested by the iterators
+implementation, but by the types these iterators return and that is provided by
+the generic parameters.
```

Comment:
Motivation can be used a precedence for future RFCs so I want this clarified to state that we are not setting a precedence to generally avoid returning concrete types in the stdlib -- which would be a mistake in my opinion. This reasoning to use opaque types due to client disinterest does not extend beyond this RFC -- specifically we are stating that we have those expectations of disinterest for iterator implementations.

Reply:
+1

Reply:
The objective of the RFC is not about concrete type vs virtual ones, but more about applying [Liskov principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle). And it's difficult to say that a principle applies only on a special RFC... My initial questioning was about only about `Array` iterators and now it extends to all collection iterators, where we have to define what collection means (because collection data structures are not limited to the collection package).

Reply:
I would vote against the RFC on those grounds. If you want to include it in the motivation you can, but I won't be in support in that case.